File: _archive/nodejs-mcp/src/tools/pdf-tools.ts
import { z } from 'zod';
import { zodToJsonSchema } from 'zod-to-json-schema';
import * as fs from 'fs';
import * as path from 'path';
import { exec } from 'child_process';
import { promisify } from 'util';

const execAsync = promisify(exec);

interface Tool {
  name: string;
  description: string;
  inputSchema: z.ZodSchema;
  handler: (args: any) => Promise<any>;
}

export class PDFTools {
  private tools: Tool[] = [];

  constructor() {
    this.initializeTools();
  }

  private initializeTools() {
    this.tools = [
      {
        name: 'view_pdf',
        description: 'Extract and view text content from a PDF file with comprehensive information including metadata, page count, and text preview',
        inputSchema: z.object({
          filePath: z.string().min(1, 'File path is required'),
          maxPages: z.number().min(1).max(100).default(10).optional(),
          includeMetadata: z.boolean().default(true).optional(),
          includeTextPreview: z.boolean().default(true).optional(),
        }),
        handler: this.viewPDF.bind(this),
      },
    ];
  }

  getTools() {
    return this.tools.map(tool => ({
      name: tool.name,
      description: tool.description,
      inputSchema: zodToJsonSchema(tool.inputSchema),
    }));
  }

  hasTool(name: string): boolean {
    return this.tools.some(tool => tool.name === name);
  }

  async handleTool(name: string, args: any) {
    const tool = this.tools.find(t => t.name === name);
    if (!tool) {
      throw new Error(`Tool ${name} not found`);
    }

    const validatedArgs = tool.inputSchema.parse(args);
    return await tool.handler(validatedArgs);
  }

  private async viewPDF(args: {
    filePath: string;
    maxPages?: number;
    includeMetadata?: boolean;
    includeTextPreview?: boolean;
  }) {
    try {
      // Check if file exists
      if (!fs.existsSync(args.filePath)) {
        throw new Error(`PDF file not found: ${args.filePath}`);
      }

      // Check if file is a PDF
      const ext = path.extname(args.filePath).toLowerCase();
      if (ext !== '.pdf') {
        throw new Error(`File is not a PDF: ${args.filePath}`);
      }

      // Get file stats
      const stats = fs.statSync(args.filePath);
      
      // Try to extract text using pdftotext (if available)
      let textContent = '';
      let pageCount = 0;
      let metadata: any = {};
      let textPreview = '';

      try {
        // Try pdftotext command
        const { stdout } = await execAsync(`pdftotext "${args.filePath}" -`);
        textContent = stdout;
        
        // Get page count
        const { stdout: pageCountOutput } = await execAsync(`pdfinfo "${args.filePath}" | grep Pages | awk '{print $2}'`);
        pageCount = parseInt(pageCountOutput.trim()) || 0;
        
        // Get text preview if requested
        if (args.includeTextPreview) {
          textPreview = textContent.substring(0, 500) + (textContent.length > 500 ? '...' : '');
        }
        
        // Get metadata if requested
        if (args.includeMetadata) {
          try {
            const { stdout: infoOutput } = await execAsync(`pdfinfo "${args.filePath}"`);
            const infoLines = infoOutput.split('\n');
            metadata = {};
            infoLines.forEach(line => {
              if (line.includes(':')) {
                const [key, value] = line.split(':', 2);
                const cleanKey = key.trim().replace(/\s+/g, '');
                metadata[cleanKey] = value.trim();
              }
            });
          } catch (metaError) {
            // Metadata extraction failed, continue without it
            console.warn('Metadata extraction failed:', metaError);
          }
        }
      } catch (cmdError) {
        // Fallback: try to read as binary and extract basic info
        textContent = `PDF file detected but text extraction failed. File size: ${this.formatFileSize(stats.size)}`;
        pageCount = 0;
        textPreview = 'Text preview not available';
        metadata = {
          note: 'Text extraction failed - pdftotext command not available or failed',
          fileSize: this.formatFileSize(stats.size),
        };
      }

      const result = {
        success: true,
        filePath: args.filePath,
        fileInfo: {
          size: stats.size,
          sizeFormatted: this.formatFileSize(stats.size),
          created: stats.birthtime,
          modified: stats.mtime,
        },
        pdfInfo: {
          pageCount: pageCount,
          ...metadata,
        },
        content: textContent,
        textPreview: args.includeTextPreview ? textPreview : undefined,
        extractedPages: Math.min(pageCount, args.maxPages || 10),
        totalPages: pageCount,
        method: 'pdftotext',
      };

      return {
        content: [
          {
            type: 'text',
            text: JSON.stringify(result, null, 2),
          },
        ],
      };
    } catch (error) {
      throw new Error(`PDF viewing failed: ${error instanceof Error ? error.message : String(error)}`);
    }
  }


  private formatFileSize(bytes: number): string {
    const sizes = ['Bytes', 'KB', 'MB', 'GB'];
    if (bytes === 0) return '0 Bytes';
    const i = Math.floor(Math.log(bytes) / Math.log(1024));
    return Math.round(bytes / Math.pow(1024, i) * 100) / 100 + ' ' + sizes[i];
  }
}

File: mcp/src/pdf_tools.py
"""
PDF text extraction tools for OneAgent MCP server.
Migrated from TypeScript pdf-tools.ts with simplified functionality.
"""

import os
import json
from typing import Optional
from datetime import datetime
import pypdf

def validate_file_path(path: str) -> bool:
    """Check if file exists and is readable."""
    return os.path.exists(path) and os.path.isfile(path)

def format_file_size(bytes: int) -> str:
    """Format file size in human readable format."""
    sizes = ['Bytes', 'KB', 'MB', 'GB']
    if bytes == 0:
        return '0 Bytes'
    i = 0
    while bytes >= 1024 and i < len(sizes) - 1:
        bytes /= 1024.0
        i += 1
    return f"{bytes:.2f} {sizes[i]}"

def extract_pdf_text(file_path: str, max_pages: int = 10) -> dict:
    """Extract text from PDF file using pypdf."""
    try:
        with open(file_path, 'rb') as file:
            pdf_reader = pypdf.PdfReader(file)
            
            # Get page count
            page_count = len(pdf_reader.pages)
            
            # Extract text from pages (limited by max_pages)
            text_content = ""
            pages_to_extract = min(page_count, max_pages)
            
            for page_num in range(pages_to_extract):
                page = pdf_reader.pages[page_num]
                text_content += page.extract_text() + "\n"
            
            # Get metadata if available
            metadata = {}
            if pdf_reader.metadata:
                metadata = {
                    'title': pdf_reader.metadata.get('/Title', ''),
                    'author': pdf_reader.metadata.get('/Author', ''),
                    'subject': pdf_reader.metadata.get('/Subject', ''),
                    'creator': pdf_reader.metadata.get('/Creator', ''),
                    'producer': pdf_reader.metadata.get('/Producer', ''),
                    'creation_date': str(pdf_reader.metadata.get('/CreationDate', '')),
                    'modification_date': str(pdf_reader.metadata.get('/ModDate', ''))
                }
            
            return {
                'success': True,
                'text_content': text_content,
                'page_count': page_count,
                'extracted_pages': pages_to_extract,
                'metadata': metadata,
                'method': 'pypdf'
            }
            
    except Exception as e:
        return {
            'success': False,
            'error': str(e),
            'method': 'pypdf'
        }

def register_pdf_tools(mcp):
    """Register PDF text extraction tools with FastMCP."""
    
    @mcp.tool()
    def view_pdf(
        file_path: str,
        max_pages: int = 10,
        include_metadata: bool = True,
        include_text_preview: bool = True
    ) -> str:
        """
        Extract and view text content from a PDF file with comprehensive information 
        including metadata, page count, and text preview.
        
        Args:
            file_path: Path to the PDF file
            max_pages: Maximum number of pages to extract (default: 10, max: 100)
            include_metadata: Whether to include PDF metadata (default: True)
            include_text_preview: Whether to include text preview (default: True)
        """
        try:
            # Validate file path
            if not validate_file_path(file_path):
                return f"Error: PDF file not found: {file_path}"
            
            # Check if file is a PDF
            if not file_path.lower().endswith('.pdf'):
                return f"Error: File is not a PDF: {file_path}"
            
            # Validate max_pages parameter
            if max_pages < 1 or max_pages > 100:
                return "Error: max_pages must be between 1 and 100"
            
            # Get file stats
            stats = os.stat(file_path)
            
            # Extract text from PDF
            extraction_result = extract_pdf_text(file_path, max_pages)
            
            if not extraction_result['success']:
                # Fallback: return basic file info if extraction fails
                result = {
                    'success': False,
                    'filePath': file_path,
                    'fileInfo': {
                        'size': stats.st_size,
                        'sizeFormatted': format_file_size(stats.st_size),
                        'created': datetime.fromtimestamp(stats.st_ctime).isoformat(),
                        'modified': datetime.fromtimestamp(stats.st_mtime).isoformat(),
                    },
                    'pdfInfo': {
                        'pageCount': 0,
                        'note': 'Text extraction failed - pypdf could not process the file'
                    },
                    'content': f"PDF file detected but text extraction failed. File size: {format_file_size(stats.st_size)}",
                    'textPreview': 'Text preview not available',
                    'extractedPages': 0,
                    'totalPages': 0,
                    'method': 'pypdf (failed)',
                    'error': extraction_result['error']
                }
            else:
                # Successful extraction
                text_content = extraction_result['text_content']
                page_count = extraction_result['page_count']
                extracted_pages = extraction_result['extracted_pages']
                
                # Create text preview if requested
                text_preview = ""
                if include_text_preview and text_content:
                    text_preview = text_content[:500] + ("..." if len(text_content) > 500 else "")
                
                result = {
                    'success': True,
                    'filePath': file_path,
                    'fileInfo': {
                        'size': stats.st_size,
                        'sizeFormatted': format_file_size(stats.st_size),
                        'created': datetime.fromtimestamp(stats.st_ctime).isoformat(),
                        'modified': datetime.fromtimestamp(stats.st_mtime).isoformat(),
                    },
                    'pdfInfo': {
                        'pageCount': page_count,
                        'extractedPages': extracted_pages,
                        'totalPages': page_count
                    },
                    'content': text_content,
                    'method': extraction_result['method']
                }
                
                # Add metadata if requested and available
                if include_metadata and extraction_result.get('metadata'):
                    result['pdfInfo']['metadata'] = extraction_result['metadata']
                
                # Add text preview if requested
                if include_text_preview:
                    result['textPreview'] = text_preview
            
            return json.dumps(result, indent=2)
            
        except Exception as e:
            return f"Error viewing PDF: {str(e)}"

# Export individual function for testing
def view_pdf(
    file_path: str,
    max_pages: int = 10,
    include_metadata: bool = True,
    include_text_preview: bool = True
) -> str:
    """
    Extract and view text content from a PDF file with comprehensive information 
    including metadata, page count, and text preview.
    
    Args:
        file_path: Path to the PDF file
        max_pages: Maximum number of pages to extract (default: 10, max: 100)
        include_metadata: Whether to include PDF metadata (default: True)
        include_text_preview: Whether to include text preview (default: True)
    """
    try:
        # Validate file path
        if not validate_file_path(file_path):
            return f"Error: PDF file not found: {file_path}"
        
        # Check if file is a PDF
        if not file_path.lower().endswith('.pdf'):
            return f"Error: File is not a PDF: {file_path}"
        
        # Validate max_pages parameter
        if max_pages < 1 or max_pages > 100:
            return "Error: max_pages must be between 1 and 100"
        
        # Get file stats
        stats = os.stat(file_path)
        
        # Extract text from PDF
        extraction_result = extract_pdf_text(file_path, max_pages)
        
        if not extraction_result['success']:
            # Fallback: return basic file info if extraction fails
            result = {
                'success': False,
                'filePath': file_path,
                'fileInfo': {
                    'size': stats.st_size,
                    'sizeFormatted': format_file_size(stats.st_size),
                    'created': datetime.fromtimestamp(stats.st_ctime).isoformat(),
                    'modified': datetime.fromtimestamp(stats.st_mtime).isoformat(),
                },
                'pdfInfo': {
                    'pageCount': 0,
                    'note': 'Text extraction failed - PyPDF2 could not process the file'
                },
                'content': f"PDF file detected but text extraction failed. File size: {format_file_size(stats.st_size)}",
                'textPreview': 'Text preview not available',
                'extractedPages': 0,
                'totalPages': 0,
                'method': 'PyPDF2 (failed)',
                'error': extraction_result['error']
            }
        else:
            # Successful extraction
            text_content = extraction_result['text_content']
            page_count = extraction_result['page_count']
            extracted_pages = extraction_result['extracted_pages']
            
            # Create text preview if requested
            text_preview = ""
            if include_text_preview and text_content:
                text_preview = text_content[:500] + ("..." if len(text_content) > 500 else "")
            
            result = {
                'success': True,
                'filePath': file_path,
                'fileInfo': {
                    'size': stats.st_size,
                    'sizeFormatted': format_file_size(stats.st_size),
                    'created': datetime.fromtimestamp(stats.st_ctime).isoformat(),
                    'modified': datetime.fromtimestamp(stats.st_mtime).isoformat(),
                },
                'pdfInfo': {
                    'pageCount': page_count,
                    'extractedPages': extracted_pages,
                    'totalPages': page_count
                },
                'content': text_content,
                'method': extraction_result['method']
            }
            
            # Add metadata if requested and available
            if include_metadata and extraction_result.get('metadata'):
                result['pdfInfo']['metadata'] = extraction_result['metadata']
            
            # Add text preview if requested
            if include_text_preview:
                result['textPreview'] = text_preview
        
        return json.dumps(result, indent=2)
        
    except Exception as e:
        return f"Error viewing PDF: {str(e)}"
